### ISSUE TITLE
Ignore labels BEFORE softmax_log

### ISSUE THREAD

**Issue by Tzawa** [chainer/chainer#1027]
The "ignore labels" function in softmax_cross_entropy ignores labels after calculating the exponents. It is also nice to add an option to ignore the labels BEFORE calculating the exponents.

**Comment by unnonouno** [comment]
Is it for efficiency?

**Comment by Tzawa** [comment]
Yes and No.

Suppose the input of softmax is (1, -1, 2), it is normalised as (e^1/Z, e^(-1)/Z, e^(2)/Z)
Here, I want to ignore the 3rd item for the normalisation as if it is -inf.
(1, -1, -inf) -> (e^1/Z, e^(-1)/Z, 0)

Of course I can realise this by artificially changing the values to be ignored to -inf, but I don't know if this is reasonable or not...

**Comment by unnonouno** [comment]
Oh, you are right, and this is a bug i think.

**Comment by unnonouno** [comment]
I found it's not true.
Shape of `x` is `(B, L)` and shape of `t` is `(B,)`, where `B` is mini-batch size and `L` is the count of labels.
This loss function is defined as `sum_i (softmax(x[i]))[t[i]]`, where `t[i]` is not equal to the ignored label.
